### PR TITLE
(merge?) lot of over simplifications of the backend

### LINKED ISSUE #261: ENH simple unifications
Unifications that seem to be within easy reach are

- **complex magnitude computation:** This is a pointwise operation that doesn't care about the shape of the array
- **nd-fourier:** with an `fftn` type operation, it should be possible to unify fourier by simply specifying the axes that we want the transform to happen on for each different package
- ~~**subsampling:** as long as we are talking about strict diadic subsampling, this can happen independently per axis, both in Fourier and in signal space;~~ CF DISCUSSION IN SLACK 

The way this should be done is to have a folder `backend_common` on the same level as `scattering1d, scattering2d, scattering3d` which implements the general functions. These are then imported into the backends specific to the scattering dimensionality. This procedure gives us the opportunity to unify bit by bit, and also to have the dimensionality-specific backends perform the necessary dimensionality-specific modifications if necessary.

### LINKED ISSUE #444: garbage error message resulting from type checks
image

"Type of A is not the Type of B" ? What is that supposed to mean for a user who just tried to call scattering on a `DoubleTensor`. We need a better error message here. Probably one stating that the input "probably needs to be of dtype float"

### PR DESCRIPTION
**Pull request by edouardoyallon**
the goal of this PR is to avoid redundant codes, in particular to uniformize the error messages, nomenclatures... this should be pretty straightforward and will probably require several rebasing

Closes #444 (the request about vacuous error messages)
Closes #261

### CODE REVIEW

File: kymatio/backend/torch_backend.py [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+def _iscomplex(x):
+    return x.shape[-1] == 2
+
+def _isreal(x):
+    return x.shape[-1] == 1
+
+
+def sanity_check(x):
+    if not _iscomplex(x):
+        raise TypeError('The input should be complex (e.g. last dimension is 2).')
```
Comment (lostanlen) [inline kymatio/backend/torch_backend.py]: e.g. ? did you mean i.e. ?

**Comment by edouardoyallon** [comment]
it'd be good if we could review it now before I dig more ..

File: kymatio/backend/tensorflow_backend.py [hunk 2]
```diff
@@ -0,0 +1,28 @@
+import numpy as np
+import tensorflow as tf
+
+def _iscomplex(x):
+    return x.dtype == np.complex64 or x.dtype == np.complex128
+
+def _isreal(x):
+    return x.dtype == np.float32 or x.dtype == np.float64
+
+class Modulus(object):
```
Comment (eickenberg) [inline kymatio/backend/tensorflow_backend.py]: do we need to inherit from `object`?

File: kymatio/backend/torch_backend.py [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+            Complex torch tensor.
+
+        Returns
+        -------
+        output : tensor
+            A tensor with the same dimensions as x, such that output[..., 0]
+            contains the complex modulus of x, while output[..., 1] = 0.
+    """
+    def __call__(self, x):
+        sanity_check(x)
```
Comment (eickenberg) [inline kymatio/backend/torch_backend.py]: lol, what is this?

File: kymatio/backend/base_backend.py [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+        self.ifft = ifft
+        self.irfft = irfft
+        self.sanity_checks = sanity_checks
+
+    def fft_forward(self, x, direction='C2C', inverse=False):
+        """Interface with torch FFT routines for 2D signals.
+
+            Example
+            -------
+            x = torch.randn(128, 32, 32, 2)
```
Comment (eickenberg) [inline kymatio/backend/base_backend.py]: This is very general, so add some more examples with the other backends

File: kymatio/backend/base_backend.py [hunk 5]
```diff
@@ -0,0 +1,62 @@
+
+
+class FFT:
+    def __init__(self, fft, ifft, irfft, sanity_checks):
```
Comment (eickenberg) [inline kymatio/backend/base_backend.py]: what is this `sanity_checks`

File: kymatio/backend/tensorflow_backend.py [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+    """This class implements a modulus transform for complex numbers.
+
+    Parameters
+    ----------
+    x: input complex tensor.
+
+    Returns
+    ----------
+    output: a real tensor equal to the modulus of x.
```
Comment (eickenberg) [inline kymatio/backend/tensorflow_backend.py]: this is wrong, cf the type case below

File: kymatio/backend/torch_backend.py [hunk 7]
```diff
@@ -0,0 +1,215 @@
...
+
+        # Special case at 0 where we return a subgradient containing 0
+        grad_input.masked_fill_(output == 0, 0)
+
+        return grad_input
+
+# shortcut for ModulusStable.apply
+modulus = ModulusStable.apply
+
+class Modulus(object):
```
Comment (eickenberg) [inline kymatio/backend/torch_backend.py]: don't inherit from object

File: kymatio/backend/torch_skcuda_backend.py [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+        raise TypeError('Input and filter must be CUDA tensors.')
+
+    if A.device.index != B.device.index:
+        raise TypeError('Input and filter must be on the same GPU.')
+
+    if _isreal(B):
+        if inplace:
+            return A.mul_(B)
+        else:
+            return A * B
```
Comment (eickenberg) [inline kymatio/backend/torch_skcuda_backend.py]: I would assume that in this case we should call `dgmm`, otherwise the skcuda backend won't be using skcuda for real multiplication

File: kymatio/scattering1d/filter_bank.py [hunk 9]
```diff
@@ -1,8 +1,7 @@
 import numpy as np
 import math
 import warnings
-from .backend.numpy_backend import ifft1d_c2c as ifft
-
+from scipy.fftpack import ifft
```
Comment (eickenberg) [inline kymatio/scattering1d/filter_bank.py]: sure about this? what's the motivation

File: kymatio/scattering2d/backend/torch_backend.py [hunk 10]
```diff
@@ -326,7 +154,10 @@ def concatenate(arrays):
 backend.cdgmm = cdgmm
 backend.modulus = Modulus()
 backend.subsample_fourier = SubsampleFourier()
-backend.fft = fft
+backend.fft = FFT(lambda x:torch.fft(x, 2, normalized=False),
+                  lambda x:torch.ifft(x, 2, normalized=False),
+                  lambda x:torch.irfft(x, 2, normalized=False, onesided=False),
+                  sanity_check)
```
Comment (eickenberg) [inline kymatio/scattering2d/backend/torch_backend.py]: In the other backend you pass `lambda x: None`. So which one is it?

File: kymatio/backend/torch_backend.py (hunk 3 above)
Comment (edouardoyallon) [reply kymatio/backend/torch_backend.py]: to remove the checks that we have done in 3 (1/2/3-D) x 4 (backends) x 2 (functions FFT, cdgmm) = 24 places

File: kymatio/backend/tensorflow_backend.py (hunk 2 above)
Comment (janden) [reply kymatio/backend/tensorflow_backend.py]: No.

File: kymatio/backend/base_backend.py (hunk 4 above)
Comment (edouardoyallon) [reply kymatio/backend/base_backend.py]: ok

File: kymatio/backend/torch_skcuda_backend.py (hunk 8 above)
Comment (edouardoyallon) [reply kymatio/backend/torch_skcuda_backend.py]: it is because torch uses dgmm for real multiplication

File: kymatio/scattering1d/filter_bank.py (hunk 9 above)
Comment (edouardoyallon) [reply kymatio/scattering1d/filter_bank.py]: I don't want the filters to be correlated with the backend

File: kymatio/scattering2d/backend/torch_backend.py (hunk 10 above)
Comment (edouardoyallon) [reply kymatio/scattering2d/backend/torch_backend.py]: i just didn't write any tests

File: kymatio/backend/torch_backend.py [hunk 11]
```diff
@@ -0,0 +1,237 @@
...
+    else:
+        C = A.new(A.shape)
+
+        A_r = A[..., 0].view(-1, B.nelement()//2)
+        A_i = A[..., 1].view(-1, B.nelement()//2)
+
+        B_r = B[..., 0].view(-1).unsqueeze(0).expand_as(A_r)
+        B_i = B[..., 1].view(-1).unsqueeze(0).expand_as(A_i)
+
+        C[..., 0].view(-1, B.nelement()//2)[:] = A_r * B_r - A_i * B_i
```
Comment (janden) [inline kymatio/backend/torch_backend.py]: Spaces around `//`.

File: kymatio/backend/torch_backend.py [hunk 12]
```diff
@@ -0,0 +1,237 @@
...
+    ----------
+    x : tensor
+        A complex tensor (that is, whose last dimension is equal to 2).
+
+    Returns
+    -------
+    x_real : tensor
+        The tensor x[..., 0] which is interpreted as the real part of x.
+    """
+    return x[..., 0]
```
Comment (janden) [inline kymatio/backend/torch_backend.py]: Newline.

Comment (janden) [inline kymatio/backend/torch_backend.py]: Newline.

File: kymatio/backend/torch_skcuda_backend.py (hunk 8 above)
Comment (janden) [reply kymatio/backend/torch_skcuda_backend.py]: Yeah we did some speed tests before and this turned out to be faster.

File: kymatio/scattering1d/backend/numpy_backend.py [hunk 13]
```diff
@@ -184,14 +119,17 @@ def concatenate(arrays):
...
+backend.modulus_complex = modulus
 backend.subsample_fourier = subsample_fourier
 backend.real = real
 backend.unpad = unpad
 backend.pad = pad
 backend.pad_1d = pad_1d
-backend.fft1d_c2c = fft1d_c2c
-backend.ifft1d_c2c = ifft1d_c2c
+backend.cdgmm = cdgmm
+backend.fft = FFT(lambda x:scipy.fftpack.fft(x),
```
Comment (janden) [inline kymatio/scattering1d/backend/numpy_backend.py]: This is weird. Why not handled by `FFT` directly?

File: kymatio/backend/numpy_backend.py [hunk 14]
```diff
@@ -0,0 +1,79 @@
...
+        x: input complex tensor.
+
+        Returns
+        -------
+        output: a real tensor equal to the modulus of x.
+
+    """
+    return np.abs(x)
+
+def _iscomplex(x):
```
Comment (janden) [inline kymatio/backend/numpy_backend.py]: `_is_complex`.

File: kymatio/backend/tensorflow_backend.py [hunk 15]
```diff
@@ -0,0 +1,83 @@
+import numpy as np
+import tensorflow as tf
+
+def _iscomplex(x):
```
Comment (janden) [inline kymatio/backend/tensorflow_backend.py]: `_Is_complex`.

File: kymatio/backend/tensorflow_backend.py [hunk 16]
```diff
@@ -0,0 +1,83 @@
...
+    return x.dtype == np.complex64 or x.dtype == np.complex128
+
+def _isreal(x):
```
Comment (janden) [inline kymatio/backend/tensorflow_backend.py]: `_is_real`.

File: kymatio/backend/numpy_backend.py [hunk 17]
```diff
@@ -0,0 +1,79 @@
...
+    return x.dtype == np.complex64 or x.dtype == np.complex128
+
+
+def _isreal(x):
```
Comment (janden) [inline kymatio/backend/numpy_backend.py]: `_is_real`.

File: kymatio/backend/numpy_backend.py [hunk 18]
```diff
@@ -0,0 +1,79 @@
...
+
+        Returns
+        -------
+        output: a real tensor equal to the modulus of x.
+
+    """
+    return np.abs(x)
+
+def _iscomplex(x):
+    return x.dtype == np.complex64 or x.dtype == np.complex128
```
Comment (janden) [inline kymatio/backend/numpy_backend.py]: Parentheses.

File: kymatio/backend/torch_backend.py [hunk 19]
```diff
@@ -0,0 +1,237 @@
+from torch.autograd import Function
+import torch
+
+BACKEND_NAME = 'torch'
+
+def _iscomplex(x):
```
Comment (janden) [inline kymatio/backend/torch_backend.py]: `_is_complex`.

File: kymatio/backend/torch_backend.py [hunk 20]
```diff
@@ -0,0 +1,237 @@
...
+    return x.shape[-1] == 2
+
+def _isreal(x):
```
Comment (janden) [inline kymatio/backend/torch_backend.py]: `_is_real`.

File: kymatio/backend/torch_skcuda_backend.py [hunk 21]
```diff
@@ -0,0 +1,83 @@
+BACKEND_NAME = 'torch'
+
+def _iscomplex(x):
```
Comment (janden) [inline kymatio/backend/torch_skcuda_backend.py]: `_is_complex`.

File: kymatio/backend/torch_skcuda_backend.py [hunk 22]
```diff
@@ -0,0 +1,83 @@
...
+    return x.shape[-1] == 2
+
+def _isreal(x):
```
Comment (janden) [inline kymatio/backend/torch_skcuda_backend.py]: `_is_real`.

File: kymatio/scattering1d/backend/torch_skcuda_backend.py [hunk 23]
```diff
@@ -232,6 +231,5 @@ def subsample_fourier(x, k):
 backend.unpad = unpad
 backend.pad = pad
 backend.pad_1d = pad_1d
-backend.fft1d_c2c = fft1d_c2c
-backend.ifft1d_c2c = ifft1d_c2c
+backend.fft = fft
 backend.concatenate = concatenate
```
Comment (janden) [inline kymatio/scattering1d/backend/torch_skcuda_backend.py]: Missing `cdgmm`.

File: kymatio/scattering1d/frontend/torch_frontend.py [hunk 24]
```diff
@@ -28,23 +28,23 @@ def register_filters(self):
             if type(k) != str:
                 # view(-1, 1).repeat(1, 2) because real numbers!
                 self.phi_f[k] = torch.from_numpy(
-                    self.phi_f[k]).float().view(-1, 1).repeat(1, 2)
+                    self.phi_f[k]).float().view(-1, 1)
```
Comment (janden) [inline kymatio/scattering1d/frontend/torch_frontend.py]: :heart:

File: kymatio/scattering1d/tests/test_torch_scattering1d.py [hunk 25]
```diff
@@ -494,18 +494,6 @@ def test_modulus(device, backend, random_state=42):
     x_grad = x2 / x_abs2[..., 0].unsqueeze(dim=-1)
     assert torch.allclose(x.grad, x_grad)
 
-    # Manually check `forward`/`backward` using fake context object. This
```
Comment (janden) [inline kymatio/scattering1d/tests/test_torch_scattering1d.py]: Why is this gone?

File: kymatio/scattering2d/backend/tensorflow_backend.py [hunk 26]
```diff
@@ -78,100 +76,18 @@ def __call__(self, x, k):
...
-backend.fft = fft
+backend.fft = FFT(lambda x: tf.signal.fft2d(x, name='ifft2d'),
+                  lambda x: tf.signal.ifft2d(x, name='ifft2d'),
+                  lambda x: tf.math.real(tf.signal.ifft2d(x, name='irfft2d')),
+                  lambda x: None)
 backend.Pad = Pad
 backend.unpad = unpad
-backend.concatenate = concatenate
+backend.concatenate = lambda x: concatenate(x, -3)
+
```
Comment (janden) [inline kymatio/scattering2d/backend/tensorflow_backend.py]: No blank line.

File: kymatio/scattering2d/tests/test_numpy_scattering2d.py [hunk 27]
```diff
@@ -116,11 +116,6 @@ def test_cdgmm_exceptions(self, backend):
                           np.empty((4, 5)).astype(np.complex128))
         assert 'first input must be complex' in record.value.args[0]
 
-        with pytest.raises(RuntimeError) as record:
```
Comment (janden) [inline kymatio/scattering2d/tests/test_numpy_scattering2d.py]: Please explain briefly why these are gone.

File: kymatio/tests/test_torch_backend.py [hunk 28]
```diff
@@ -0,0 +1,25 @@
...
+
+    class FakeContext:
+        def save_for_backward(self, *args):
+            self.saved_tensors = args
+
+    ctx = FakeContext()
+    y = ModulusStable.forward(ctx, x)
+    y_grad = torch.ones_like(y)
+    x_grad_manual = ModulusStable.backward(ctx, y_grad)
+    assert torch.allclose(x_grad_manual, x_grad)
```
Comment (janden) [inline kymatio/tests/test_torch_backend.py]: Newline.

File: kymatio/tests/test_torch_backend.py [hunk 29]
```diff
@@ -0,0 +1,25 @@
+import torch
```
Comment (janden) [inline kymatio/tests/test_torch_backend.py]: I assume there will be others for NP and TF?

File: kymatio/scattering1d/tests/test_torch_scattering1d.py (hunk 25 above)
Comment (janden) [reply kymatio/scattering1d/tests/test_torch_scattering1d.py]: Nevermind.

**Comment by janden** [comment]
Moves to #536.
